File: stream/static/js/events.js
events = {}
events.handlers = {}

events.register_handler = function( name, func ) 
{
	if( !(name in events.handlers) )
		events.handlers[name] = [];

	events.handlers[name].push(func);
}

events.handle_event = function( msg )
{
	event = JSON.parse(msg);
	events.fire( event.name, event.payload );
	
}

events.fire = function(name, payload)
{
	if( !(name in events.handlers) )
	{
		console.log("Ignoring event '" + name + "' (no handler)");
		return;
	}

	console.log("Handling event: " + name);
	console.log(payload);
	for( var e = 0; e < events.handlers[name].length; e++ )
		events.handlers[name][e]( payload );
}

// Event fired when someone plays a track. 
//
// Expected payload:
// track_id: the id of the track that was played
// track_playcount: the new playcount for the track
// album_id: the album id the track was off of
// album_playcount: the new playcount for the album
// artist_id: the artist id of the track
// artist_playcount: the new playcount for that artist
events.register_handler( "track_played", function(p)
{
	$(".playcount[data-ps-model=track][data-ps-id=" + p.track_id + "]").text(p.track_playcount);
	$(".playcount[data-ps-model=album][data-ps-id=" + p.album_id + "]").text(p.album_playcount);
	$(".playcount[data-ps-model=artist][data-ps-id=" + p.artist_id + "]").text(p.artist_playcount);
});

events.register_handler( "authenticated", function(p)
{
	$("#pane_settings_login").text("You are logged in as " + p.username + ".");
});

events.register_handler( "pinged", function(p)
{
	console.log("Pong!");
});

File: stream/util/events.py
from ws4redis.publisher import RedisPublisher
from ws4redis.redis_store import RedisMessage
import json

event_publisher = RedisPublisher(facility="events", broadcast=True)

def publish_event( name, payload, pub = event_publisher ):
	pub.publish_message(RedisMessage(json.dumps({"name": name, "payload": payload})))

def get_allowed_channels(request, channels):
	return set(channels).intersection(['subscribe-broadcast', 'subscribe-session'])

def track_played( track ):
	event_payload = {}
	event_payload["track_id"] = track.id
	event_payload["track_playcount"] = track.playcount
	event_payload["album_id"] = track.album.id
	event_payload["album_playcount"] = track.album.playcount
	event_payload["artist_id"] = track.artist.id
	event_payload["artist_playcount"] = track.artist.playcount

	publish_event( "track_played", event_payload )

def authenticated( session_key, username ):
	session_publisher = RedisPublisher(facility="events", sessions=[session_key])
	publish_event( "authenticated", {"username": username}, session_publisher )
	
def pinged( session_key ):
	session_publisher = RedisPublisher(facility="events", sessions=[session_key])
	publish_event( "pinged", {}, session_publisher )	
